{"cells":[{"cell_type":"markdown","id":"ce6590a1","metadata":{"id":"ce6590a1"},"source":["# Assignment 08 Solutions"]},{"cell_type":"markdown","id":"efb72245","metadata":{"id":"efb72245"},"source":["#### Q1. What are the two latest user-defined exception constraints in Python 3.X?\n","**Ans:** **`raise`** and **`assert`** are the two latest user-defined exception constraints in Python 3.X"]},{"cell_type":"markdown","id":"f1479448","metadata":{"id":"f1479448"},"source":["#### Q2. How are class-based exceptions that have been raised matched to handlers?\n","**Ans:** In python, Users can define custom exceptions by creating a new class. This exception class has to be derived, either directly or indirectly from built-in **`Exception`** class. This new exception class like other exceptions can be raised using the **`raise`** statement with an optional error message."]},{"cell_type":"code","source":["class agecheckerforelct(Exception):\n","  def __init__(self,msg):\n","    self.msg=msg\n","age=int(input('enter a num'))\n","if age<=18:\n","  raise agecheckerforelct('too young to elect')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":218},"id":"7WK93WI89hq1","executionInfo":{"status":"error","timestamp":1690723421720,"user_tz":-330,"elapsed":3332,"user":{"displayName":"Suman Hachappa","userId":"02923451167331183866"}},"outputId":"4bd40a17-b3a2-4f21-98f1-e139d49ec89f"},"id":"7WK93WI89hq1","execution_count":2,"outputs":[{"name":"stdout","output_type":"stream","text":["enter a num12\n"]},{"output_type":"error","ename":"agecheckerforelct","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31magecheckerforelct\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-a4c2a096dcb6>\u001b[0m in \u001b[0;36m<cell line: 5>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mage\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'enter a num'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mage\u001b[0m\u001b[0;34m<=\u001b[0m\u001b[0;36m18\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m   \u001b[0;32mraise\u001b[0m \u001b[0magecheckerforelct\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'too young to elect'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31magecheckerforelct\u001b[0m: too young to elect"]}]},{"cell_type":"markdown","id":"15ce7b10","metadata":{"id":"15ce7b10"},"source":["#### Q3. Describe two methods for attaching context information to exception artefacts ?\n","**Ans:** The **`process()`** method of LoggerAdapter is where the contextual information is added to the logging output. its passes the message and keyword arguments of the logging call, and it passes back modified versions of these to use in the call to the underlying logger.\n","\n","Other method that can be used is **`exception()`**, Logs a messgae with level ERROR on this logger. The arguments are interpreted as for **`debug()`**. Exception info is added to the logging message."]},{"cell_type":"markdown","id":"7fbfd775","metadata":{"id":"7fbfd775"},"source":["#### Q4. Describe two methods for specifying the text of an exception object's error message ?\n","**Ans:** **`raise`** and **`assert`** are two methods for specifying the text of an exception object's error message.**`raise`** statement is used to trigger explicit exception, if certain condition is not as per requirement of programmer. it helps in triggering exception as per need of programmer and logic.\n","\n","There are few assertions that programmer always want to be True to avoid code failure. This type of requirment is fullfilled by **`assert`** statement. This statement takes a Boolean Condition output of which if True, further program executes. if output of assert statement is False it raises an **`Assertion Error`**."]},{"cell_type":"markdown","id":"0855de9b","metadata":{"id":"0855de9b"},"source":["#### Q5. Why do you no longer use string-based exceptions?\n","**Ans:** String-based Exceptions doesn't inherit from Exceptions. so plain exceptions catch all exceptions and not only system."]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.13"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}